incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -150,14 +241,45 @@ public MessageConsumer newStream(
     public void onAttached(
         long bindingId)
     {
+        MqttKafkaBindingConfig binding = supplyBinding.apply(bindingId);
+        if (willAvailable && coreIndex == 0)
+        {
+            Optional<MqttKafkaRouteConfig> route = binding.routes.stream().findFirst();
+            final long routeId = route.map(mqttKafkaRouteConfig -> mqttKafkaRouteConfig.id).orElse(0L);
+
+            willProxy = new KafkaWillProxy(binding.id, routeId,
+                binding.sessionsTopic(), binding.messagesTopic(), binding.retainedTopic());
+
+            this.reconnectAt = signaler.signalAt(
+                currentTimeMillis(),
+                SIGNAL_INITIATE_WILL_KAFKA_STREAM,
+                this::onWillKafkaStreamInitializationSignal);
+        }
         sessionIds.put(bindingId, supplySessionId.get());
     }
 
+    private void onWillKafkaStreamInitializationSignal(
+        int signalId)
+    {
+        assert signalId == SIGNAL_INITIATE_WILL_KAFKA_STREAM;
+
+        this.reconnectAt = NO_CANCEL_ID;
+
+        willProxy.doKafkaBegin(supplyTraceId.get(), 0, 0);
+    }
+
     @Override
     public void onDetached(
         long bindingId)
     {
         sessionIds.remove(bindingId);
+
+        if (willProxy != null)
+        {
+            willProxy.doKafkaEnd(supplyTraceId.get(), 0);
+            signaler.cancel(reconnectAt);
+            reconnectAt = NO_CANCEL_ID;
+        }
```

Comment:
One thing to add here: I can't use the remove in onDetached since we remove the binding in the MqttKafkaProxyFactory. This would require some restructuring of the code: moving the bindings in the individual factories like MqttKafkaSessionFactory etc.. Let me know if you think it's worth to change it.

Reply:
Other then this I applied this feedback